### ISSUE TITLE
`ragged_dot` gives nonzero grad to empty groups

### ISSUE THREAD

**Issue by renos** [openxla/tokamax#887]
`tokamax.ragged_dot`'s VJP for `rhs` is broken in two ways:

1. Groups with `group_sizes[g] == 0` get a **nonzero** gradient. They should get zero since they contributed nothing to the output.
2. All unused groups get the **same** spurious gradient value.
3. Even groups that *did* receive tokens get a gradient that doesn't match `jax.lax.ragged_dot`.

This quietly breaks MoE training: every expert head gets gradient on every step regardless of whether it was routed to, so per-expert specialization doesn't work.

## Environment

- `tokamax` 0.0.10
- `jax` 0.9.0, `jaxlib` 0.9.0
- Python 3.13.11
- NVIDIA RTX 6000 Ada Generation, CUDA

## Reproducer

```python
import jax
import jax.numpy as jnp
from tokamax import ragged_dot as tokamax_ragged_dot


def make_loss(ragged_dot_fn, num_groups):
    def loss_fn(rhs, lhs, expert_ids):
        sort_idx = jnp.argsort(expert_ids)
        sorted_lhs = lhs[sort_idx]
        sorted_ids = expert_ids[sort_idx]
        group_sizes = jnp.bincount(sorted_ids, length=num_groups)
        return ragged_dot_fn(sorted_lhs, rhs, group_sizes).sum()
    return loss_fn


num_groups, K, N, M = 18, 64, 43, 32
rng = jax.random.PRNGKey(0)
rhs = jax.random.normal(rng, (num_groups, K, N))
lhs = jax.random.normal(jax.random.PRNGKey(1), (M, K))

# Tokens distributed across groups 0–5; groups 6–17 receive ZERO tokens.
expert_ids = jnp.array(
    [0, 0, 1, 1, 2, 3, 3, 3, 3, 4, 4, 5, 5, 5, 5, 5,
     0, 0, 1, 1, 2, 3, 3, 3, 3, 4, 4, 5, 5, 5, 5, 5],
    dtype=jnp.int32,
)

for name, fn in [("jax.lax.ragged_dot", jax.lax.ragged_dot),
                 ("tokamax.ragged_dot", tokamax_ragged_dot)]:
    g = jax.grad(make_loss(fn, num_groups))(rhs, lhs, expert_ids)
    norms = jnp.linalg.norm(g, axis=(1, 2))
    print(f"\n{name}:")
    for i in range(num_groups):
        print(f"  group {i:2d}: {float(norms[i]):.4e}")
```

## Output

```
jax.lax.ragged_dot:
  group  0: 1.0105e+02
  group  1: 1.1208e+02
  group  2: 7.1685e+01
  group  3: 1.3538e+02
  group  4: 1.0271e+02
  group  5: 1.4602e+02
  group  6: 0.0000e+00          <-- correct: group 6 received 0 tokens
  group  7: 0.0000e+00
  ...
  group 17: 0.0000e+00

tokamax.ragged_dot:
  group  0: 1.0105e+02
  group  1: 1.1208e+02
  group  2: 7.1685e+01
  group  3: 1.3538e+02
  group  4: 1.0271e+02
  group  5: 1.4602e+02
  group  6: 2.6241e+02          <-- wrong: should be 0
  group  7: 2.6241e+02          <-- same value across all unused groups
  ...
  group 17: 2.6241e+02
```

If we route zero tokens to group 3 instead (assign those 8 slots to other groups), `jax.lax.ragged_dot` correctly returns `0` for group 3, while `tokamax.ragged_dot` returns ~`1.8e+03` for group 3 and the same `2.6e+02` for the other unused groups.

**Comment by renos** [comment]
Fix proposed in #889
